src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -138,80 +139,99 @@ public override AssemblySymbol ContainingAssembly
+                if (_lazyHasIsUnmanagedAttribute.Value() && (_lazyFlags & (GenericParameterAttributes.ReferenceTypeConstraint | GenericParameterAttributes.DefaultConstructorConstraint)) != 0)
                 {
-                    var constraint = metadataReader.GetGenericParameterConstraint(constraintHandle);
-                    var constraintTypeHandle = constraint.Type;
+                    _lazyFlags = _lazyFlags & ~(GenericParameterAttributes.ReferenceTypeConstraint | GenericParameterAttributes.DefaultConstructorConstraint);
+                    _lazyConstraintsUseSiteErrorInfo = new CSDiagnosticInfo(ErrorCode.ERR_BindToBogus, this);
+                }
 
-                    TypeSymbol typeSymbol = tokenDecoder.GetTypeOfToken(constraintTypeHandle);
+                if (constraints.Count > 0)
+                {
+                    var symbolsBuilder = ArrayBuilder<TypeSymbol>.GetInstance();
+                    MetadataDecoder tokenDecoder;
 
-                    // Drop 'System.Object' constraint type.
-                    if (typeSymbol.SpecialType == SpecialType.System_Object)
+                    if ((object)containingMethod != null)
                     {
-                        continue;
+                        tokenDecoder = new MetadataDecoder(moduleSymbol, containingMethod);
                     }
-
-                    // Drop 'System.ValueType' constraint type if the 'valuetype' constraint was also specified.
-                    if (((_flags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0) &&
-                        (typeSymbol.SpecialType == SpecialType.System_ValueType))
+                    else
                     {
-                        continue;
+                        tokenDecoder = new MetadataDecoder(moduleSymbol, containingType);
                     }
 
-                    typeSymbol = TupleTypeDecoder.DecodeTupleTypesIfApplicable(typeSymbol,
-                                                                               constraintHandle,
-                                                                               moduleSymbol);
+                    foreach (var constraintHandle in constraints)
+                    {
+                        var constraint = metadataReader.GetGenericParameterConstraint(constraintHandle);
+                        var typeSymbol = tokenDecoder.DecodeGenericParameterConstraint(constraint.Type, out bool hasUnmanagedModreq);
 
-                    symbolsBuilder.Add(typeSymbol);
+                        if (hasUnmanagedModreq != this._lazyHasIsUnmanagedAttribute.Value())
+                        {
+                            // The presence of UnmanagedType modreq has to match the presence of the IsUnmanagedAttribute
+                            Interlocked.CompareExchange(ref _lazyConstraintsUseSiteErrorInfo, new CSDiagnosticInfo(ErrorCode.ERR_BindToBogus, this), CSDiagnosticInfo.EmptyErrorInfo);
+                            continue;
```

Comment:
>continue [](start = 28, length = 8)

It doesn't feel right to continue here. We reported an error, but we should not ignore the constraint type that was specified in metadata.